utils/timer.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#include "timer.h"
+#include "utils/logger.h"
+
+#define TIMER_LOGGER "timer"
+
+static logger_id_t logger_id;
+static pthread_mutex_t timer_lock = PTHREAD_MUTEX_INITIALIZER;
+struct _ta_timer_t {
+  pthread_t *thread_id;
+  timer_t *timer_id;
+};
+
+void timer_logger_init() { logger_id = logger_helper_enable(TIMER_LOGGER, LOGGER_DEBUG, true); }
+
+int timer_logger_release() {
+  logger_helper_release(logger_id);
+  if (logger_helper_destroy() != RC_OK) {
+    ta_log_error("Destroying logger failed %s.\n", TIMER_LOGGER);
+    return EXIT_FAILURE;
+  }
+
+  return 0;
+}
+
+static void handler(union sigval v) {
+  ta_timer_t *args = (ta_timer_t *)v.sival_ptr;
+  ta_log_error("timer_id: %p expired\n", *(args->timer_id));
+
+  /* Cancel the thread */
+  pthread_mutex_lock(&timer_lock);
+  pthread_cancel(*(args->thread_id));
+  pthread_mutex_unlock(&timer_lock);
+}
+
+ta_timer_t *ta_timer_start(const struct itimerspec *timer, void *callback, void *args) {
+  pthread_t *thread_id = NULL;
```

Comment:
You don't have to set zero to these 3 variables initially since they are then assigned with malloc.